docs/guide/counter/plugins/counter/counter.go

```diff
@@ -1,98 +1,197 @@
-	TotalFees types.Coins
+// CounterTx
+//--------------------------------------------------------------------------------
+
+// register the tx type with it's validation logic
+// make sure to use the name of the handler as the prefix in the tx type,
+// so it gets routed properly
+const (
+	NameCounter = "cntr"
+	ByteTx      = 0x21
+	TypeTx      = NameCounter + "/count"
+)
+
+func init() {
+	basecoin.TxMapper.RegisterImplementation(CounterTx{}, TypeTx, ByteTx)
 }
 
 type CounterTx struct {
-	Valid bool
-	Fee   types.Coins
+	Valid    bool        `json:"valid"`
+	Fee      types.Coins `json:"fee"`
+	Sequence int         `json:"sequence"`
+}
+
+func NewCounterTx(valid bool, fee types.Coins, sequence int) basecoin.Tx {
+	return CounterTx{
+		Valid:    valid,
+		Fee:      fee,
+		Sequence: sequence,
+	}.Wrap()
+}
+
+func (c CounterTx) Wrap() basecoin.Tx {
+	return basecoin.Tx{c}
 }
 
+// ValidateBasic just makes sure the Fee is a valid, non-negative value
+func (c CounterTx) ValidateBasic() error {
+	if !c.Fee.IsValid() {
+		return coin.ErrInvalidCoins()
+	}
+	if !c.Fee.IsNonnegative() {
+		return coin.ErrInvalidCoins()
+	}
+	return nil
+}
+
+// Custom errors
 //--------------------------------------------------------------------------------
 
-type CounterPlugin struct {
+var (
+	errInvalidCounter = rawerr.New("Counter Tx marked invalid")
+)
+
+// This is a custom error class
+func ErrInvalidCounter() error {
+	return errors.WithCode(errInvalidCounter, abci.CodeType_BaseInvalidInput)
+}
+func IsInvalidCounterErr(err error) bool {
+	return errors.IsSameError(errInvalidCounter, err)
 }
 
-func (cp *CounterPlugin) Name() string {
-	return "counter"
+// This is just a helper function to return a generic "internal error"
+func ErrDecoding() error {
+	return errors.ErrInternal("Error decoding state")
 }
 
-func (cp *CounterPlugin) StateKey() []byte {
-	return []byte(fmt.Sprintf("CounterPlugin.State"))
+// CounterHandler
+//--------------------------------------------------------------------------------
+
+func NewCounterHandler() basecoin.Handler {
+	// use the default stack
+	coin := coin.NewHandler()
+	counter := CounterHandler{}
+	dispatcher := stack.NewDispatcher(
+		stack.WrapHandler(coin),
+		counter,
+	)
+	return stack.NewDefault().Use(dispatcher)
 }
 
-func New() *CounterPlugin {
-	return &CounterPlugin{}
+type CounterHandler struct {
+	stack.NopOption
 }
 
-func (cp *CounterPlugin) SetOption(store types.KVStore, key, value string) (log string) {
-	return ""
+var _ stack.Dispatchable = CounterHandler{}
+
+func (_ CounterHandler) Name() string {
+	return NameCounter
 }
 
-func (cp *CounterPlugin) RunTx(store types.KVStore, ctx types.CallContext, txBytes []byte) (res abci.Result) {
-	// Decode tx
-	var tx CounterTx
-	err := wire.ReadBinaryBytes(txBytes, &tx)
-	if err != nil {
-		return abci.ErrBaseEncodingError.AppendLog("Error decoding tx: " + err.Error()).PrependLog("CounterTx Error: ")
-	}
+func (_ CounterHandler) AssertDispatcher() {}
```

Comment:
Really?  What does it like?

I prefer to no bind to a variable name to clarify that this is a sort of class-method that doesn't depend on struct state.  Is there a better way to do this?